x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/WatcherLifeCycleService.java

```diff
@@ -99,8 +104,18 @@ public void clusterChanged(ClusterChangedEvent event) {
         if (isWatcherStoppedManually) {
             if (this.state.get() == WatcherState.STARTED) {
                 clearAllocationIds();
-                watcherService.stop("watcher manually marked to shutdown by cluster state update");
-                this.state.set(WatcherState.STOPPED);
+                this.state.compareAndSet(WatcherState.STARTED, WatcherState.STOPPING);
+                //waiting to set state to stopped until after all currently running watches are finished
+                watcherService.stop("watcher manually marked to shutdown by cluster state update", () -> {
```

Comment:
can this be added inside an if statement?

```java
boolean stopping = this.state.compareAndSet(WatcherState.STARTED, WatcherState.STOPPING);
if (stopping) {
 ...
}
```

This way we know `WatcherService.stop(...)` will only be invoked once.

Reply:
good call, updated.